lightstep/trace_context.py

```diff
@@ -0,0 +1,311 @@
+            r"[{0}\d][-{0}\d_*/]{{0,240}}@[{0}][-{0}\d_*/]{{0,13}}"
+        ).format("a-z"),
+        # https://www.w3.org/TR/2019/CR-trace-context-20190813/#value
+        r"[ {0}]{{0,255}}[{0}]".format(
+            escape(
+                r"".join(
+                    [
+                        chr(character) for character in range(33, 127)
+                        if character not in [44, 61]
+                    ]
+                )
+            )
+        )
+    )
+)
+
+_BLANK = re_compile(r"^\s*$")
+
+
+class TraceContextPropagator(Propagator):
+    """
+    Propagator for the W3C Trace Context format
+
+    The latest version of this Candidate Recommendation can be found here:
+    https://www.w3.org/TR/trace-context/
+
+    This is an implementation for this specific version:
+    https://www.w3.org/TR/2019/CR-trace-context-20190813/
+
+    The Candidate Recommendation will be referred to as "the document" in the
+    comments of this implementation. This implementation adds comments which
+    are URLs that point to the specific part of the document that explain the
+    rationale of the code that follows the comment.
+
+    There is a set of test cases defined for testing any implementation of the
+    document. This set of test cases can be found here:
+    https://github.com/w3c/trace-context/tree/master/test
+
+    This set of test cases will be referred to as "the test suite" in the
+    comments of this implementation.
+    """
+
+    def inject(self, span_context, carrier):
+
+        carrier[_TRACEPARENT] = "00-{}-{}-{}".format(
+            format(span_context.trace_id, "032x"),
+            format(span_context.span_id, "016x"),
+            format(span_context.baggage.pop("trace-flags", 0), "02x")
+        )
+
+        carrier.update(span_context.baggage)
+
+    def extract(self, carrier):
+        traceparent = None
+        tracestate = None
+
+        multiple_header_template = "Found more than one header value for {}"
+
+        # FIXME Define more specific exceptions and use them instead of the
+        # generic ones that are being raised below
+
+        # https://www.w3.org/TR/2019/CR-trace-context-20190813/#header-name
+
+        trace_context_free_carrier = {}
+
+        for key, value in carrier.items():
+
+            lower_key = key.lower()
+
+            # The document requires that the headers be accepted regardless of
+            # the case of their characters. This means that the keys of carrier
+            # may contain 2 or more strings that match traceparent or
+            # tracestate when the case of the characters of such strings is
+            # ignored. The document does not specify what is to be done in such
+            # a situation, this implementation will raise an exception.
+            if lower_key == _TRACEPARENT:
+
+                if traceparent is None:
+                    traceparent = value
+
+                else:
+                    raise Exception(
```

Comment:
What about `SpanContextCorruptedException`?

Reply:
And same for the errors below (I know you put a FIXME, but I think it should be straightforward to replace almost all of them with `SpanContextCorruptedException` ;) )

Reply:
Fixed!